src/sas/sascalc/fit/fitstate.py

```diff
@@ -0,0 +1,407 @@
+    #print("pars", model.params)
+    #print("limits", model.details)
+    #print("fitted", fitted)
+
+    # Set the resolution
+    data = copy.deepcopy(state.data)
+    if state.disable_smearer:
+        smearer = None
+    elif state.enable_smearer:
+        smearer = smear_selection(data, model)
+    elif state.pinhole_smearer:
+        # see sasgui/perspectives/fitting/basepage.py: reset_page_helper
+        dx_percent = state.dx_percent
+        if state.dx_old:
+            dx_percent = 100*(state.dx_percent / data.x[0])
+        # see sasgui/perspectives/fitting/fitpage.py: _set_pinhole_smear
+        percent = dx_percent / 100.
+        if state.enable2D:
+            # smear_type is Pinhole2D.
+            # dqx_data, dqy_data initialized to 0
+            data.dqx_data = percent * data.qx_data
+            data.dqy_data = percent * data.qy_data
+        else:
+            data.dx = percent * data.x
+            data.dxl = data.dxw = None  # be sure it is not slit-smeared
+        smearer = smear_selection(data, model)
+    elif state.slit_smearer:
+        # see sasgui/perspectives/fitting/fitpage.py: _set_pinhole_smear
+        data_len = len(data.x)
+        data.dx = None
+        data.dxl = (state.dxl if state.dxl is not None else 0.) * np.ones(data_len)
+        data.dxw = (state.dxw if state.dxw is not None else 0.) * np.ones(data_len)
+        smearer = smear_selection(data, model)
+    else:
+        raise ValueError("expected resolution specification for fit")
+
+    # Set the data weighting (dI, sqrt(I), I, or uniform)
+    weight = get_data_weight(state)
+
+    # Make fit data object and set the data weights
+    # TODO: check 2D masked data
+    if state.enable2D:
+        fitdata = FitData2D(sas_data2d=data, data=data.data,
+                            err_data=data.err_data)
+        fitdata.err_data = weight
+    else:
+        data.mask = (np.isnan(data.y) if data.y is not None
+                        else np.zeros_like(data.x, dtype='bool'))
+        fitdata = FitData1D(x=data.x, y=data.y,
+                            dx=data.dx, dy=data.dy, smearer=smearer)
+        fitdata.dy = weight
+        fitdata.set_fit_range(qmin=state.qmin, qmax=state.qmax)
+    fitdata.sas_data = data
+
+    # Don't need initial values since they have been stuffed into the model
+    # If provided, then they should be one-to-one with the parameter names
+    # listed in fitted.
+    initial_values = None
+
+    fitmodel = Model(model, fitdata)
+    fitmodel.name = state.fit_page
+    fitness = SasFitness(
+        model=fitmodel,
+        data=fitdata,
+        constraints=state.constraints,
+        fitted=fitted,
+        initial_values=initial_values,
+        )
+
+    return fitness
+
+class BumpsPlugin:
+    """
+    Object holding methods for interacting with SasView using the direct
+    bumps interface.
+    """
+    #@staticmethod
+    #def data_view():
+    #    pass
+
+    #@staticmethod
+    #def model_view():
+    #    pass
+
```

Comment:
Why staticmethod is used here? I just cannot see it but there is probably a good reason.

Reply:
`BumpsPlugin` is defining a namespace for a set of functions which is passed to `bumps.cli.install_plugin` (line 393 below).  `install_plugin` is expecting a module, and using `getattr` to retrieve particular named functions from the module.  A set of `staticmethod` in a class is duck-type equivalent.